text/0019-beforeSendTransaction.md

```diff
@@ -0,0 +1,102 @@
+* Start Date: 2022-09-26
+* RFC Type: feature
+* RFC PR: https://github.com/getsentry/rfcs/pull/19
+
+# Summary
+
+Proposal to make `beforeSend` (or a similar method) work for transactions (and possibly other payload types in the future) in all SDKs.
+
+# Motivation
+
+For error events, `beforeSend` allows the user to make whatever changes they want to event data and to filter events however they want, which is useful for solving the cases where our built-in mechanisms for doing so don't correctly handle their particular use case. There is no such hook for transactions. Instead, users have to use `addGlobalEventProcessor` to add an event processor to the scope. This is less than ideal in a few ways:
+
+- Ease of use (or lack thereof)
+  - Nothing about `beforeSend` is obviously errors-only, so users assume it applies everywhere and try to use it for transactions.
+  - Users have to learn two different ways to do effectively the same thing.
+  - It forces users to write code in two separate places, which leads to potential redundancy if they want to act on any properties shared by error and transaction events.
+
+- Lack of ultimate control over transaction events
+  - Event processors happen in an unspecified order, and there's no guarantee a processor added by a user will run last. This means that an integration, for example, might change event data after the user's last chance to intervene.
+
+(That last point is not _strictly_ true. As a last resort, a user could create a custom transport or could proxy events through their own servers, but both of those are pretty ugly, cumbersome workarounds.)
+
+# Background
+
+When transactions were introduced, the decision was made not to run them through `beforeSend` because they follow a slightly different schema, and therefore had the potential to break any existing `beforeSend` which relies on its input being a certain shape. It's unclear whether a transaction-specific `beforeSend`-type hook was discussed at the time.
```

Comment:
> If I remember correctly, there we good reasons for not calling beforeSend for transactions

Yes - from what I recall, it was exactly the reason I listed there, that the schema mis-match had the potential to break people's `beforeSend` functions. Or do you mean there was a reason for not having a transaction-specific `beforeSend`-like option?

Reply:
> It's unclear whether a transaction-specific `beforeSend`-type hook was discussed at the time.

I think this sentence is inaccurate. We discussed it and most likely the outcome was what you just pointed out: `the schema mis-match had the potential to break people's beforeSend functions`

> Or do you mean there was a reason for not having a transaction-specific beforeSend-like option

Could also be, but I can't recall that.

Reply:
Sorry, I think we're miscommunicating here, because I think we actually agree on the history. Let me try to rephrase:

I know that back in the day, when tracing was brand new, we discussed sending transactions through `beforeSend` itself, and decided against it because of the schema issue potentially creating a breaking change. What I don't know, and you're saying you can't recall either, is if a _separate_ `beforeSend`-like option (`beforeSendTransaction` or similar) was considered, and if it was, why it wasn't implemented. With regard to that question schemas/breaking changes are irrelevant, because no one would have had any existing code to break.

Does that make sense?

Reply:
>  if a _separate_ beforeSend-like option (beforeSendTransaction or similar) was considered, and if it was, why it wasn't implemented. With regard to that question schemas/breaking changes are irrelevant, because no one would have had any existing code to break.

Yep, that makes sense 👍